sbin/pfdhcplistener

```diff
@@ -343,10 +360,214 @@ sub process_pkt {
 }
 
 sub normal_sighandler {
-    Net::Pcap::pcap_breakloop($pcap);
+    $running = 0;
+    if ($pcap) {
+        Net::Pcap::pcap_breakloop($pcap);
+    }
     $logger->trace( "pfdhcplistener: caught SIG" . $_[0] . " - terminating" );
 }
 
+=head2 runtasks
+
+run all runtasks
+
+=cut
+
+sub runtasks {
+    my $mask = POSIX::SigSet->new(POSIX::SIGCHLD());
+    sigprocmask(SIG_BLOCK, $mask);
+    while (@REGISTERED_TASKS) {
+        my $task = shift @REGISTERED_TASKS;
+        runtask($task);
+    }
+    sigprocmask(SIG_UNBLOCK, $mask);
+}
+
+=head2 runtask
+
+creates a new child to run a task
+
+=cut
+
+sub runtask {
+    my ($task) = @_;
+    my $pid = fork();
+    if ($pid) {
+        $CHILDREN{$pid} = $task;
+    }
+    elsif ($pid == 0) {
+        $SIG{CHLD} = 'DEFAULT';
+        $IS_CHILD = 1;
+        Log::Log4perl::MDC->put('tid', $$);
+        setup_global($task);
+        reload_config();
+        _runtask($interface);
+    }
+    else {
+    }
+}
+
+=head2 _runtask
+
+the task to is ran in a loop until it is finished
+
+=cut
+
+sub _runtask {
+    my ($interface) = @_;
+    $0 = "$0 - $interface";
+    setup_pcap($interface);
+    while (is_running()) {
+        my $result = Net::Pcap::dispatch( $pcap, -1, \&process_pkt, [ $interface , $pcap ] );
+        $logger->logdie(Net::Pcap::pcap_geterr($pcap)) if ($result == -1);
+    }
+    exit;
+}
+
+=head2 is_running
+
+is_running
+
+=cut
+
+sub is_running {
+    return $running && is_parent_alive();
+}
+
+=head2 is_parent_alive
+
+Checks to see if parent is alive
+
+=cut
+
+sub is_parent_alive {
+    kill(0, $PARENT_PID);
+}
+
+=head2 register_task
+
+registers the task to run
+
+=cut
+
+sub register_task {
+    my ($taskId) = @_;
+    push @REGISTERED_TASKS, $taskId;
+
+}
+
+=head2 waitforit
+
+waits for signals
+
+=cut
+
+sub waitforit {
+    while ($running) {
+        pause;
+        runtasks();
+    }
+}
+
+=head2 alarm_sighandler
+
+the alarm signal handler
+
+=cut
+
+sub alarm_sighandler {
+    $ALARM_RECV = 1;
+}
+
+=head2 child_sighandler
+
+reaps the children
+
+=cut
+
+sub child_sighandler {
+    local ($!, $?);
+    while (1) {
+        my $child = waitpid(-1, WNOHANG);
+        last unless $child > 0;
+        my $task = delete $CHILDREN{$child};
+        register_task($task);
+    }
+}
+
+=head2 registertasks
+
+    Register all tasks
+
+=cut
+
+sub registertasks {
+    for my $task (get_tasks()) {
+        register_task($task);
```

Comment:
minor nitpick here, but can we have a better name than register_task?
even _doRegisterTask would be better, or else it's confusing to have two subs with pretty much the same name.